### DISCUSSION TITLE
Is There A Chance That MicroPython will get as good as C++?

### DISCUSSION THREAD

**Discussion by ovexro** [micropython/micropython#15807] [Core Development]
Hello.

I hope this question doesn't offend anyone, but Claude compares C++ and MicroPython and says that MicroPython must catch up to C++.

Is there a chance that MicroPython will get as good as C++?

Thanks.

**Comment by robert-hh** [comment]
> Is there a chance that MicroPython will get as good as C++?

The questions seems nonsense to me. In my ears it sounds like: Will apples get the same as oranges?

Besides that: what do you call good? Python and C++ are different, and both are good at their playgrounds. MicroPython is a  restricted version of Python, it is not a replacement for code written in C++.

**Reply by ovexro** [reply]
Thank you, Robert! I was confused as to what I should pick between C++ and MicroPython. Because Claude and ChatGPT recommend C++ instead of MicroPython; both say that MicroPython is not that good.

**Reply by stinos** [reply]
They're not human, they lack the skills to point out the only corect reasoning (i.e., Robert's) :)

**Reply by GitHubsSilverBullet** [reply]
> Thank you, Robert! I was confused as to what I should pick between C++ and MicroPython. Because Claude and ChatGPT recommend C++ instead of MicroPython; both say that MicroPython is not that good.

If you believe anything any LLM throws at you, you've completely lost control of your life!

**Reply by peterhinch** [reply]
rofl!

**Reply by sosi-deadeye** [reply]
The question is about: Development Speed vs. Execution Speed

Micropython: A dynamic language with a bytecode compiler and interpreter. If you know python very well, you can develop very fast. Execution speed could not compare with C/C++. Optimization could be done with decorators (`micropython.native`, `micropython.viper`) or with assembler. I think you can also use compiled mpy-Files, which includes different languages, but this afik. in the making.

C/C++: Static complied language, runs on everything, runs fast and let you shoot yourself in the foot.

**Reply by peterhinch** [reply]
As one who's coded in both languages, I know which I prefer. Life offers more enjoyable experiences than hours spent hunting down that last memory leak in an a big application.

**Reply by sk8board** [reply]
In general, which is faster, [viper](https://github.com/micropython/micropython/wiki/Improving-performance-with-Viper-code) or [Native machine code in .mpy files](https://docs.micropython.org/en/latest/develop/natmod.html#natmod) or [MicroPython external C modules](https://docs.micropython.org/en/latest/develop/cmodules.html)?

**Reply by jonnor** [reply]
Native code is faster than any bytecode, as long as FFI overheads are kept small (meaning a sizable chunk of work is done per function invocation). Dynamic module or C module can be quite similar. Though .mpy files execute from RAM - whereas custom C modules execute from FLASH. Which can have different performance characteristics - hardware dependent.
But the best is to do some benchmarks for the workloads you are interested in. And share the results :)

**Reply by peterhinch** [reply]
The phrase "Native machine code" is perhaps ambiguous. Carefully crafted code using the inline assembler is extremely fast. By contrast the `native` decorator often doesn't make much difference (in my experience). With care and the "right" kind of code the `viper` decorator can be **fast** -  comparable to C or Assembler. The "right" kind of code being bit manipulations, addressing of arrays through pointers, integer arithmetic etc.

This is all hand-waving, albeit based on experience and measurement. I don't know of any generally accepted quantitative results.
